### ISSUE TITLE
How to make selector wait for API content to appear on screen?

### ISSUE THREAD

**Issue by liamlows** [elrumordelaluz/reactour#515]
Howdy,

First off great package, I love the design and experience that comes with it and made adding this feature super easy. 

However, I am running into and issue and I'm not sure if its due to my ignorance or if it is a limitation of the package. I am using Next.JS and on one of my tours of 9 steps, on step 5 i need to navigate to another page for two steps then return back to the previous page. I have managed to navigate in Next.JS using `actionAfter` with a `router.push('/page2')`. I have also used a custom setCurrentStep as can be seen in this sandbox link (https://codesandbox.io/s/friendly-shape-3f90tq).

The issue im experiencing is the on the new page i am navigating to, I am using URL parameters to conditionally render a specific component on that page, but that component uses API calls to populate data and conditionally render content. Due to this when the navigation occurs, the walk through cannot initially find the selector and chooses a small location to select in the top left corner of DOM. After a few milliseconds, the content appears on the page and the selector does not update. The same issue occurs when coming back to the other page as it has dynamic content as well and fails to load.

Is there a way to delay the selector until the api content renders and then have it select that content? I have thought about creating additional steps to select container elements that have shape and will soon contain data and then moving to the api content, but if a user uses the dots to navigate the problem would still exist.

Any ideas?

**Comment by liamlows** [comment]
Also to add another comment on this issue, one thing we have had to do is initialize the tour provider in the root of our app with an empty array for the steps. We found we must do this as we have different tours for each page and each page uses a useEffect with the setSteps hook to initialize the steps for that page. We check if a tour isOpen (using useTour) before setting the steps to avoid setting steps during a tour and it also waits for an api call to finish to do so (we are using SWR so there is not a promise we can access related to those calls, just data variables and isLoading/isError variables that update as data comes in).

I created another sandbox to mimic the current use of tour in our application more closely. Unfortunately, in the sandbox, all works fine.

I also saw a note stating to not use actions to fire off navigation changes. Due to the nature of our applications (multiple tours with the tour provider only being able to be defined at the root) we cannot use the method documented by react tour where setCurrentStep is provided to the tour provider.

Im really not sure why the tour acts so buggy when we navigate between pages and often times fails to highlight the correct element (in the action the element parameter passes null). Does a portal perhaps need to be used?

**Comment by liamlows** [comment]
Ok so I managed to solve my issue by getting the right mix of selectors, mutationObservables, highlightedSelectors, and resizeObservables. I think some more documentation could be useful here as to what exactly each one does and the criteria needed for them to trigger since I read what was in the docs, but that did not help me solve my problem. 

Regardless though my point still stands as to the best way to handle asynchronous data and multiple tours. Should we only have one tour provider and set steps to null initially but update them as needed per tour? Can we have multiple providers? Is there a way to have a provider and tour set up on a single page and then have another on a different page to isolate issues with overlap?

@elrumordelaluz should I move this to a new issue and resolve this one? Let me know, I am happy to build out docs more as needed to cover some of this newer info if we decide to go that route. Thanks again for all the effort thus far!

**Comment by kWh61** [comment]
@liamlows I'm encountering the same problem. Could you perhaps share how you were able to find a workaround for this? I saw in your first codesandbox link that asynchronous actions work when going from step 1 to step 2 (but not from step 2 to step 1 for some reason). Appreciate the help!

**Comment by liamlows** [comment]
@kWh61 so the codesandbox show it not working from step 2 to 1, that was the issue i had. I found that it is likely due to the fact that my steps have actions that make calls to router to move between pages. 

The "workaround" to avoid issues is to set your selector to a static element that will always be in screen on initial page load and then set the mutationObservables, highlightedSelectors, and resizeObservables to the element that will conditionally render. Doing this for the specific instance i was having issues with fixed the problem.

Another way is to add an additional step before to select the base page element which will give the asynchronous data time to load and then when the user hits next it will correctly move to the element. This wouldn't work if you allow your users to jump around the tour via dot navigation.

Hoping one of the maintainers/contributors may have some advice, fairly certain this is a documentation misunderstanding on my part lol.

**Comment by MarcLopezAvila** [comment]
Probably kind of related to https://github.com/elrumordelaluz/reactour/issues/487

I think mutationObservables, highlightedSelectors, and resizeObservables should be improved a bit to support a more flexible configuration and asynchrony.

**Comment by jt6677** [comment]
Is there any solution like fallback if selector is not present?  I lazyload my components. So when I redirect from one page to the other, there will be some delay. How can I elegantly make the TourProvider wait for component fully loading?

**Comment by jt6677** [comment]
I thought this type of async action is common.

**Comment by jt6677** [comment]
I wasted entire day trying to get around this problem. It hurts me to type this. I should just call it sunk cost and use a different lib.

**Comment by elrumordelaluz** [comment]
Hi all and first of all sorry for the pretty late reply.
> I wasted entire day trying to get around this problem. It hurts me to type this. I should just call it sunk cost and use a different lib.

I am sorry that you had to switch lib and I agree that is a common use-case what you are describing.

Let me switch this issue in some points to then analyse if proceed in separated issues/threads or just here with your help, who are using the lib in real life projects.

- Shadow DOM and selectors for _Observables_ (sorry again @MarcLopezAvila that didn't had time to create a reproduction here)
- Since the _Observables_ cannot be the only way to solve the async updates in steps, analyse options here like event emitters or similar approaches.
- _Observables_ documentation it's a standalone thing, needs to be explained well in plain english and probably with more than one example/use-case. Probably should be updated/improved the documentation in genereal.

Let me know what do you think and thank you for your help!

**Comment by elrumordelaluz** [comment]
Thank you for your feedback @jt6677!

**Comment by kazuyoshi-yamada-line** [comment]
I have the same issue. So I'm thinking about solving this problem and putting PR in it. I am planning to solve this problem by putting the disableWhenSelectorFalsy flag on tour or step.

What do you think about this idea? @elrumordelaluz

**Comment by elrumordelaluz** [comment]
> I am planning to solve this problem by putting the disableWhenSelectorFalsy flag on tour props or stepType.

Sounds pretty good @kazuyoshi-yamada-line!

Waiting for your contribution and thank you!

**Comment by kazuyoshi-yamada-line** [comment]
I posted Pull Request ( #546 ) about this issue.

**Comment by ConstantinChirila** [comment]
@liamlows So how did you solve this in the end. Been banging my head on this for the past few days.

**Comment by elrumordelaluz** [comment]
Merged, available at `@reactour/tour@3.5.0`.

**Comment by MarcLopezAvila** [comment]
> Merged, available at `@reactour/tour@3.5.0`.

Hi, is this release also fixing the issue when using shadow dom?

Thanks
